(odsp-client): Add odsp-client as experimental package

sonalideshpandemsft opened the pull request:
## Description

This PR introduces a new client that provides a set of common utility functions to be used whenever developers want to create and load Fluid containers backed by files stored on ODSP. The scope of the current package is marked fluid-experimental and the apis are marked with the alpha release tag. 

## Reviewer Guidance

Would appreciate feedback on:
- API names
- Coding patterns
- Implementation details
- Return values of functions

## Other information or known dependencies

[MSFT Internal Design Doc](https://microsoft-my.sharepoint-df.com/:w:/p/sdeshpande/EbuksgepVV9HqS-RuPaXI2ABJtsO1aZIoBLwpSEdrUj25w?e=0m1dOR)

[AB#6224](https://dev.azure.com/fluidframework/internal/_workitems/edit/6224)

sonalideshpandemsft committed:
init odsp client

sonalideshpandemsft committed:
move to service-client

sonalideshpandemsft committed:
audience

Josmithr commented on packages/service-client/odsp-client/.eslintrc.js:
```diff
@@ -0,0 +1,32 @@
+ */
+
+module.exports = {
+	extends: [require.resolve("@fluidframework/eslint-config-fluid/minimal"), "prettier"],
```
We should start with the recommended config at least, rather than the minimal one.
```suggestion
	extends: [require.resolve("@fluidframework/eslint-config-fluid"), "prettier"],
```

Josmithr commented on packages/service-client/odsp-client/.eslintrc.js:
```diff
@@ -0,0 +1,32 @@
+		// This library is used in the browser, so we don't want dependencies on most node libraries.
+		"import/no-nodejs-modules": ["error", { allow: ["events"] }],
+	},
+	overrides: [
```
Do we need these overrides?

Josmithr commented on packages/service-client/odsp-client/.gitignore:
```diff
@@ -0,0 +1,52 @@
+# Compiled TypeScript and CSS
```
We can probably minimize this. The repo-root gitignore should cover most of this.

Josmithr commented on packages/service-client/odsp-client/package.json:
```diff
@@ -0,0 +1,90 @@
+{
```
Let's be sure to add a README in this package explaining what it does, how to use it, etc. (preferably with examples).

Josmithr commented on packages/service-client/odsp-client/src/OdspAudience.ts:
```diff
@@ -0,0 +1,26 @@
+/*!
```
Note: our other modules use `camelCase` file naming, rather than `PascalCase` naming. We should be consistent with this.

Josmithr commented on packages/service-client/odsp-client/src/index.ts:
```diff
@@ -0,0 +1,15 @@
+	OdspMember,
+} from "./interfaces";
+export { OdspClient } from "./OdspClient";
+export { OdspAudience } from "./OdspAudience";
```
Do we need to export the class, or is just the interface sufficient? Do external users need to construct these?

Josmithr commented on packages/service-client/odsp-client/src/interfaces.ts:
```diff
@@ -0,0 +1,118 @@
+import { type IUser } from "@fluidframework/protocol-definitions";
+
+/**
+ * OdspConnectionConfig defines the necessary properties that will be applied to all containers
```
Nit
```suggestion
 * Defines the necessary properties that will be applied to all containers
```

Josmithr commented on packages/service-client/odsp-client/src/interfaces.ts:
```diff
@@ -0,0 +1,118 @@
+ */
+export interface OdspConnectionConfig {
+	/**
+	 * Site url representing ODSP resource location
```
Are there more concrete docs we can point to here so users know how to provide this?

Josmithr commented on packages/service-client/odsp-client/src/interfaces.ts:
```diff
@@ -0,0 +1,118 @@
+	 */
+	driveId: string;
+
+	path: string;
```
Docs? 😁

Josmithr commented on packages/service-client/odsp-client/src/OdspAudience.ts:
```diff
@@ -0,0 +1,26 @@
+export class OdspAudience extends ServiceAudience<OdspMember> implements IOdspAudience {
+	protected createServiceMember(audienceMember: IClient): OdspMember {
+		const user = audienceMember.user as OdspUser;
+		assert(user?.name !== undefined, 'Provided user was not an "OdspUser".');
```
It looks like `email` is also a required property of `OdspUser`. We should verify it exists here too.

Josmithr commented on packages/service-client/odsp-client/src/OdspClient.ts:
```diff
@@ -0,0 +1,201 @@
+import { OdspAudience } from "./OdspAudience";
+
+/**
+ * @alpha
```
Do we expect external users to subclass this at all? If not, we should annotate this as `@sealed`.

Josmithr commented on packages/service-client/odsp-client/src/interfaces.ts:
```diff
@@ -0,0 +1,118 @@
+}
+
+/**
+ * This interface holds attributes specific to the odsp service
```
Nit
```suggestion
 * This interface holds attributes specific to the ODSP service
```

Josmithr commented on packages/service-client/odsp-client/api-report/odsp-client.api.md:
```diff
@@ -0,0 +1,83 @@
+import { ITokenProvider } from '@fluidframework/azure-client';
+import { ServiceAudience } from '@fluidframework/fluid-static';
+
+// @alpha (undocumented)
```
I know this is still a draft, but ideally any of the items currently marked as `undocumented` should be documented before we merge. More details on these items may also help reviewers evaluate the API surface.

sonalideshpandemsft committed:
fix ci failures and some feedback

sonalideshpandemsft committed:
fix ci failures

sonalideshpandemsft committed:
add docs

sonalideshpandemsft committed:
add readme

ChumpChief commented on packages/service-client/odsp-client/package.json:
Let's mark this with `private: true` for now too, until we're ready for broader exposure.

sonalideshpandemsft committed:
add docs

sonalideshpandemsft committed:
feedback

ChumpChief commented on packages/service-client/odsp-client/src/odspClient.ts:
```diff
@@ -0,0 +1,205 @@
+			 * a new `itemId` is created in the user's drive, which developers can use for various operations
+			 * like updating, renaming, moving the Fluid file, changing permissions, and more.
+			 */
+			return resolvedUrl.itemId;
```
We can look at it in a followup, but the return value from attach should be whatever we can then feed into `getContainer()` to retrieve the container (also see other note above about wanting `getContainer()` to take an id rather than a sharingUrl).

sonalideshpandemsft replied:
Returning the absolute URL instead of the item id.

ChumpChief commented on packages/service-client/odsp-client/src/odspClient.ts:
```diff
@@ -0,0 +1,205 @@
+		const getAttributes = async (): Promise<OdspServiceAttributes> => {
+			const resolvedUrl = container.resolvedUrl as IOdspResolvedUrl;
+			if (resolvedUrl === undefined) {
+				throw new Error("Resolved Url not available on attached container");
```
In general, we always expect these to be defined after attach.  This approach is fine to start with, but it would probably make sense for us to instead await attachment, if the container is not yet attached (rather than reject the promise).

sonalideshpandemsft replied:
adding this to doc feedback section

ChumpChief reviewed (approved):
This looks good for an initial checkin I think.  I'm sure there will be more iteration and feedback but definitely a great starting point.

scottn12 reviewed (commented):
Great work Sonali! Excited to see this progress 🙂

vladsud commented on layerInfo.json:
```diff
@@ -157,6 +157,17 @@
+					"Framework",
+					"Runtime",
+					"Routerlicious-Driver",
+					"Server-Shared-Utils",
```
I'd think there should be no dependencies for the last 3 items, no?

sonalideshpandemsft replied:
AzureClient will still remain for now. But rest can be removed :)

vladsud commented on packages/service-client/odsp-client/README.md:
```diff
@@ -0,0 +1,101 @@
+## @fluid-experimental/odsp-client
+
+The odsp-client package provides a simple and powerful way to consume collaborative Fluid data with the ODSP service.
```
consume data with service? Does not sound right to me. Maybe build collaborative Fluid applications with ODSP as a storage mechanism?

vladsud commented on packages/service-client/odsp-client/README.md:
```diff
@@ -0,0 +1,101 @@
+
+### Experimental Features
+
+`OdspClient` supports the ability to instantiate with experimental features enabled.
```
I think some English proofing is required :)
to instantiate what?
Maybe instantiate itself (does not sound), better - could be instantiated with experimental features

vladsud commented on packages/service-client/odsp-client/README.md:
```diff
@@ -0,0 +1,101 @@
+### Experimental Features
+
+`OdspClient` supports the ability to instantiate with experimental features enabled.
+These features are experimental in nature and should **NOT** be used in production applications.
```
I think should or should not is not our call. We only tell users what to expect / level of support

vladsud commented on packages/service-client/odsp-client/README.md:
```diff
@@ -0,0 +1,101 @@
+
+## Fluid Containers
+
+A Container instance is a organizational unit within Fluid. Each Container instance has a connection to the defined Fluid Service and contains a collection of collaborative objects.
```
I do not like that articulation. Not clear what is "connection to defined Fluid Service" (does it mean AFR or ODSP?)
Container maps directly to a file - a unit of storage, with all of its semantics (permissions), but also Fluid adds its own semantics (ordering of ops is per container).

vladsud commented on packages/service-client/odsp-client/README.md:
```diff
@@ -0,0 +1,101 @@
+
+## Defining Fluid Containers
+
+Fluid Containers are defined by a schema. The schema includes initial properties of the Container as well as what collaborative objects can be dynamically created.
```
Maybe "content / interpretation of Fluid Containers is defined...."
Fluid container is defined by its ID :)

vladsud commented on packages/service-client/odsp-client/README.md:
```diff
@@ -0,0 +1,101 @@
+	],
+};
+const OdspClient = new OdspClient(clientProps);
+const { container, services } = await OdspClient.createContainer(containerSchema);
```
So there is no way to specify file name, right?
And it's possible (for load path) to specify itemID that does not match URL & Path info in clientProps.
Both seem wrong to me.

While we might want not to expose filename concept (and just use guide), users are interacting with storage through GraphAPI as well, and might want to name their files. This might come handy for various workflows like search, eDiscovery that clients might need to deal with. I think it should be possible to specify a file name.

sonalideshpandemsft replied:
todo: will iterate on this in upcoming PRs. Possibly moving site url, drive id, file path, fileName close to container attachment.

vladsud commented on packages/service-client/odsp-client/README.md:
```diff
@@ -0,0 +1,101 @@
+};
+
+// Fetch back the container that had been created earlier with the same url and schema
+const { container, services } = await OdspClient.getContainer("_unique-url_", schema);
```
It will be hard for developers to use this flow, without schema being stored as part of the file. How would developer go by changing that schema, and supporting V1, V2, ... files? Any project have to go through it eventually, it's not the question of IF, it's a question of WHEN

vladsud commented on packages/service-client/odsp-client/src/interfaces.ts:
```diff
@@ -0,0 +1,135 @@
+ */
+export interface OdspContainerServices {
+	/**
+	 * Retrieves tenant-specific attributes associated with the ODSP service for the current Fluid container.
```
It's hard to part what "tenant-specific" attributes means here. Tenant (in context of ODSP) is Microsoft. Cisco, etc.
I'd just say - container-specific attributes.

vladsud commented on packages/service-client/odsp-client/src/interfaces.ts:
```diff
@@ -0,0 +1,135 @@
+ * Represents attributes specific to the ODSP service for a Fluid container.
+ * @alpha
+ */
+export interface OdspServiceAttributes {
```
This interface looks a bit weird to me. On one hand, it has container-speicific info (sharing URL), so name likely should reflect that (it's not about service, it's about container).
On another hand it does not contain such thing as itemID - the most useful thing for clients who need to use ODSP API to deal with this file

vladsud commented on packages/service-client/odsp-client/src/interfaces.ts:
```diff
@@ -0,0 +1,135 @@
+ * {@link @fluidframework/protocol-definitions#IMember} interface to include this service-specific value.
+ * @alpha
+ */
+export interface OdspUser extends IUser {
```
Can this be not here? I'e somewhere in ODSP driver and reexported here?

sonalideshpandemsft replied:
Will iterate on this in upcoming PRs

vladsud commented on packages/service-client/odsp-client/src/odspClient.ts:
```diff
@@ -0,0 +1,205 @@
+				token: tokenResponse.jwt,
+			};
+		};
+		this.documentServiceFactory = new OdspDocumentServiceFactory(
```
I'd love to see mistakes of AzureClient being addressed. We need a way for the client to provide their own factory / build adapters on top of it.
The moment it's out, we will tell people they can use caching capabilities built into the driver, but to do so - they need to pass policy and cache implementation to driver factory. How users will be able to do so without cloning this code?

We need to follow principle - simple things should be easy, complex things should be possible.
Or - there have to be escape path.

sonalideshpandemsft replied:
TODO: adding this as follow-up

vladsud commented on packages/service-client/odsp-client/src/odspClient.ts:
```diff
@@ -0,0 +1,205 @@
+
+	public async createContainer(containerSchema: ContainerSchema): Promise<{
+		container: IFluidContainer;
+		services: OdspContainerServices;
```
Any good reason why services is not a member of IFluidContainer? Looking at getContainerServices(), it should be a static function, and the could be on any object.

sonalideshpandemsft replied:
IFluidContainer consists of client-side properties and methods. @sashasimic @ChumpChief @Josmithr, can add more context.

vladsud commented on packages/service-client/odsp-client/src/odspClient.ts:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	public async getContainer(
+		sharingUrl: string,
```
I've left comments in document in PR about it: 
1. I'm not sure it has to be sharing link (and types of sharing links we support). It has perf impact.
2. I do not understand how it all works if URL points to different drive ID or different siteURI / path that are specified when OdspClient is created. Do we ignore them? What driveID is returned on different APIs - one that was supplied to OdspClient, or one that identifies a file?
3. I think we are confusing users about exposing too many concepts, like URI, containerID, itemID. I think we should expose one, and provide static functions to get all the other data.
4. I assume we will not allow users to provide their own URI resolvers, and leave them to do conversion of their app API -> our ID/path outside of these interactions (but it would be worth confirming and describing it somewhere)

sonalideshpandemsft replied:
Regarding (2): 
`I do not understand how it all works if URL points to different drive ID or different siteURI / path that are specified when OdspClient is created. Do we ignore them?` - I'll look into this more. 

`What driveID is returned on different APIs - one that was supplied to OdspClient, or one that identifies a file?` - Does this mean there will be two separate drive ids? Why would there be a different driveId to identify a file?
You can only use one driveId for any given Fluid Container. So, if you need to use more, you can create additional odsp-clients. But the Drive ID will be different for each tenant.

jatgarg replied:
@ChumpChief @vladsud  I checked the way azure client is doing this. It seems like that id is not enough there too. The azure endpoint is taken in constructor and then using this id we make some modifications to that endpoint and append this id to reach to the request url which is then used to load the container. I am not sure I like that a lot because that is kind of functionality provided by resolver and its utility apis. Similarly, ODSP also has endpoint encoding all this info but imo it is not a sound way to arrive at the request url. 

I don't see any harm in taking the url when app can make use of the utility api to construct it and just pass it to this function. And it does not necessarily be a sharing link, getAbsoluteUrl in OdspDriverUrlResoover does not give back a asharing link, but it is still enough to load the container as the absolute url is resolvable by the same resolver.

vladsud commented on packages/service-client/odsp-client/src/odspClient.ts:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	private createLoader(containerSchema: ContainerSchema): Loader {
+		const runtimeFactory = new DOProviderContainerRuntimeFactory(containerSchema);
```
That's how we did it in AzureClient, but that looks wrong to me. Schema should be defined by file content, not by external code. It's not very clear what happens when they do not match.

tylerbutler replied:
What do you mean "file content" defines schema?

vladsud commented on packages/service-client/odsp-client/src/odspClient.ts:
```diff
@@ -0,0 +1,205 @@
+			permission: [],
+			scopes: [],
+			user: { id: "" },
+			mode: "write",
```
This feels like should be "read" mode.
Any time file is changed causes a cascade of actions in SPO / Substrate world. It's rather expensive for us if every open would result in signals flowing through Substrate / Working set, even if that's the only thing happening (no search indexing, no new file revisions generated).

tylerbutler replied:
If it starts in read mode how does the developer switch?

sonalideshpandemsft commented on layerInfo.json:
```diff
@@ -157,6 +157,17 @@
+					"Runtime",
+					"Routerlicious-Driver",
+					"Server-Shared-Utils",
+					"AzureClient"
```
adding to doc feedback section. will iterate on this in upcoming PRs

sonalideshpandemsft commented on packages/service-client/odsp-client/package.json:
```diff
@@ -0,0 +1,92 @@
+		"temp-directory": "nyc/.nyc_output"
+	},
+	"dependencies": {
+		"@fluidframework/azure-client": "workspace:~",
```
added this to doc feedback section

sonalideshpandemsft committed:
feedback

sonalideshpandemsft committed:
feedback

sonalideshpandemsft committed:
fix ci failures

jenn-le committed:
test(tree): test concurrent restoration and edit of removed tree (#18305)

adds a test for the concurrent restoration and edit of removed tree

---------

Co-authored-by: Noah Encke <Noah.Encke@microsoft.com>

jatgarg committed:
Make improvements in calculating retry after time for calls to storage/relayService (#18306)

## Description

Make improvements in calculating retry after time for calls to
storage/relayService. The logic at service side is not very sound to
calculate this value. So at least in short term, improving this logic at
client side to be intelligent in deciding this value.
Loop faced a bug related to this, so we need to provide a short-term
viable sol.

---------

Co-authored-by: Jatin Garg <jatingarg@Jatins-MacBook-Pro-2.local>

CraigMacomber committed:
tree2: Fix build issue from merge (#18316)

## Description

Two recent PRs conflicted to break the build. This fixes it.

alexvy86 committed:
Update deps in server release group: axios to 1.6.2, compression to 4.6.2 (#18312)

CraigMacomber committed:
tree2: Remove needless import cycle (#18320)

## Description

Fix import to remove cycle.

sonalideshpandemsft commented on layerInfo.json:
```diff
@@ -157,6 +157,10 @@
 			},
+			"OdspClient": {
+				"dirs": ["packages/service-client/odsp-client"],
+				"deps": ["AzureClient"]
```
ITokenResponse is part of azure client. I have opened a work item to move this to fluid static or somewhere else. But that's only the shared piece as of now.

sonalideshpandemsft committed:
Merge remote-tracking branch 'origin' into alpha-m365-client

sonalideshpandemsft committed:
fix ci failures

sonalideshpandemsft committed:
feedback

sonalideshpandemsft commented:
I'm merging this PR to move forward. I've documented the majority of the feedback as work items in ADO.
